mono/metadata/exception.c

```diff
@@ -683,6 +683,24 @@ mono_get_exception_argument (const char *arg, const char *msg)
 	return ex;
 }
 
+TYPED_HANDLE_DECL (MonoArgumentException);
+
+MonoExceptionHandle
+mono_exception_new_argument (const char *arg, const char *msg, MonoError *error)
+{
+	MonoExceptionHandle ex;
+	ex = mono_exception_new_by_name_msg (mono_get_corlib (), "System", "ArgumentException", msg, error);
+
+	if (arg && !MONO_HANDLE_IS_NULL (ex)) {
+		MonoArgumentExceptionHandle argex = (MonoArgumentExceptionHandle)ex;
+		MonoStringHandle arg_str = mono_string_new_handle (MONO_HANDLE_DOMAIN (ex), arg, error);
+		mono_error_assert_ok (error);
```

Comment:
Why asserting here? Can't we `return_val_if_nok`?

Reply:
We have a recurring bad pattern that attempting to raise an exception can fail due to low memory. It seems non ideal and maybe watch for it?? Though "watch for it" isn't a great solution. In this case though I guess it is ok, as long as the earlier allocation succeeds. So just remove the assert?

Reply:
Another recurring pattern is raising an exception may or may not accept a MonoError. We seem to think, and I tend to agree, that you should always take a MonoError, because almost anything can fail, but that then raises the specter or what to do in error paths, which I was sort of trying to address here.  I think just return on error and not assert. Try to leave it to the next layer -- though in particular an error in exception  handling is not practical. Errors can/should be tolerated almost anywhere, except in error handling. Anyway, ok to remove it and replace with something gentler.

Reply:
Yes return_val_if_nok, but return ex or return NULL, is a subtle question. Is the exception good enough without the argument? My "strict" self says no but my "lax" self says probably. And then, if we return ex, we should clear the error?? Or maybe never set it? All very subtle and minor but I don't like to do things just by accident, whether strict or lax. Let me come back to this in a few minutes, see if there is an easy lax pattern.

Reply:
This is also what mono_get_exception_argument is doing here.

Reply:
Though it lacks the output error.

Reply:
Just removing the assert seems reasonable.
We'll then  just set the field to null, which is redundant but ok.

Question remains as to if we should actually clear the error and let the exception be considered created completely enough. Really it is the entire off-roadmap question as to low memory handling. Exceptions need some level of preallocation. But we don't like reusing them, which hurts, and they have some dynamism like here, which hurts more. A more general idea but less precise and more wasteful is have some reserve memory, like 1k or 4k or 64k per thread. 1k is probably enough.

Reply:
This was just a question. If assertion makes more sense, then let's assert :)

And the way you are doing it now just looks like you are ignoring the error - which is not the case since you are not cleaning it up - so if anything, it seems more confusing to me now.